gusto/transport_equation.py

```diff
@@ -0,0 +1,345 @@
+
+    def mass_term(self, q):
+        return inner(self.test, q)*dx
+
+    @abstractmethod
+    def advection_term(self):
+        pass
+
+
+class LinearAdvection(TransportEquation):
+    """
+    Class for linear transport equation.
+
+    :arg state: :class:`.State` object.
+    :arg V: :class:`.FunctionSpace object. The function space that q lives in.
+    :arg qbar: The reference function that the equation has been linearised
+               about. It is assumed that the reference velocity is zero and
+               the ubar below is the nonlinear advecting velocity
+               0.5*(u'^(n+1) + u'(n)))
+    :arg ibp: string, stands for 'integrate by parts' and can take the value
+              None, "once" or "twice". Defaults to "once".
+    :arg continuity: is the equation in continuity form L(q) = div(u'*qbar)?
+                     If False, the equation is assumed to be in advective
+                     form L(q) = u' dot grad(qbar)
+    """
+
+    def __init__(self, state, V, qbar, ibp=None, continuity=False):
+        super(LinearAdvection, self).__init__(state, V, ibp)
+        self.continuity = continuity
+        self.qbar = qbar
+        # currently only used with the following option combinations:
+        if continuity and ibp is not "once":
+            raise NotImplementedError("If we are solving a linear continuity equation, we integrate by parts once")
+        if not continuity and ibp is not None:
+            raise NotImplementedError("If we are solving a linear advection equation, we do not integrate by parts.")
+
+    def advection_term(self, q):
+
+        if self.continuity:
+            L = (-dot(grad(self.test), self.ubar)*self.qbar*dx +
+                 jump(self.ubar*self.test, self.n)*avg(self.qbar)*self.dS)
+        else:
+            L = self.test*dot(self.ubar,self.state.k)*dot(self.state.k,grad(self.qbar))*dx
+        return L
+
+
+class Advection(TransportEquation):
+    """
+    Class for DG discretisation of the transport equation.
+
+    :arg state: :class:`.State` object.
+    :arg V: :class:`.FunctionSpace object. The function space that q lives in.
+    :arg ibp: string, stands for 'integrate by parts' and can take the value
+              None, "once" or "twice". Defaults to "once".
+    :arg continuity: is the equation in continuity form L(q) = div(u*q)?
+                     If False, the equation is assumed to be in advective
+                     form L(q) = u dot grad(q)
+                     Defaults to False
+    """
+    def __init__(self, state, V, ibp="once", continuity=False):
+        super(Advection, self).__init__(state, V, ibp)
+        self.continuity = continuity
+
+    def advection_term(self, q):
+
+        if self.continuity:
+            if self.ibp is "once":
+                L = -inner(grad(self.test), outer(q, self.ubar))*dx
+            else:
+                L = inner(self.test, div(outer(q, self.ubar)))*dx
```

Comment:
This branch can never be hit, right?  Because in the superclass constructor you ensure that ibp is `once` if continuity is True.